test/categorical_algebra/DPO.jl

```diff
@@ -0,0 +1,738 @@
+
+IWD = WiringDiagram([], []);
+fv, gv = add_box!(IWD, Box(:f,[],[])), add_box!(IWD, Box(:g,[],[]));
+
+XWD = WiringDiagram([:A,:A],[:D,:B])
+fv, gv, hv = add_box!(XWD, f), add_box!(XWD, g), add_box!(XWD, h)
+fv2, iv = add_box!(XWD, f), add_box!(XWD, i);
+add_wires!(XWD, Pair[
+  (input_id(XWD),1) => (fv,1),
+  (iv,1) => (iv, 1), (fv, 1) => (iv, 2), (iv, 2)=>(gv,1),
+  (gv,1) => (hv,1),
+  (hv,1) => (output_id(XWD),1),
+  (input_id(XWD),2) => (fv2,1),
+  (fv2,1) => (output_id(XWD),2)]);
+
+L=ACSetTransformation(IWD.diagram, LWD.diagram, Box=[1,2]);
+R=ACSetTransformation(IWD.diagram, RWD.diagram, Box=[1,3]);
+m=ACSetTransformation(LWD.diagram, GWD.diagram, Box=[1,2],InPort=[2],OutPort=[1],Wire=[1]);
+@test is_isomorphic(XWD.diagram, rewrite_match(L,R,m))
+# to turn result back into a wiring diagram:
+# result = WiringDiagram([],[]); result.diagram = rewrite_match(L,R,m)
+
+
+
+
+
+######################################################################################
+######################################################################################
+# Petri Nets
+######################################################################################
+######################################################################################
+vectorify(n::Vector) = n
+vectorify(n::Tuple) = length(n) == 1 ? [n] : n
+vectorify(n) = [n]
+state_dict(n) = Dict(s=>i for (i, s) in enumerate(n))
+
+# Petri Nets
+############
+
+@present TheoryPetriNet(FreeSchema) begin
+  (T, S, I, O) :: Ob
+  it::Hom(I,T)
+  is::Hom(I,S)
+  ot::Hom(O,T)
+  os::Hom(O,S)
+end
+
+const AbstractPetriNet = AbstractACSetType(TheoryPetriNet)
+
+add_species!(p::AbstractPetriNet;kw...) = add_part!(p,:S;kw...)
+add_species!(p::AbstractPetriNet,n;kw...) = add_parts!(p,:S,n;kw...)
+add_transition!(p::AbstractPetriNet;kw...) = add_part!(p,:T;kw...)
+add_transitions!(p::AbstractPetriNet,n;kw...) = add_parts!(p,:T,n;kw...)
+add_input!(p::AbstractPetriNet,t,s;kw...) = add_part!(p,:I;it=t,is=s,kw...)
+add_inputs!(p::AbstractPetriNet,n,t,s;kw...) = add_parts!(p,:I,n;it=t,is=s,kw...)
+add_output!(p::AbstractPetriNet,t,s;kw...) = add_part!(p,:O;ot=t,os=s,kw...)
+add_outputs!(p::AbstractPetriNet,n,t,s;kw...) = add_parts!(p,:O,n;ot=t,os=s,kw...)
+sname(p::AbstractPetriNet,s) = (1:ns(p))[s]
+tname(p::AbstractPetriNet,t) = (1:nt(p))[t]
+snames(p::AbstractPetriNet) = map(s->sname(p, s), 1:ns(p))
+tnames(p::AbstractPetriNet) = map(t->tname(p, t), 1:nt(p))
+
+@present TheoryLabelledReactionNet <: TheoryPetriNet begin
+  (Rate, Concentration, Name)::Data
+  rate::Attr(T, Rate)
+  concentration::Attr(S, Concentration)
+  tname::Attr(T, Name)
+  sname::Attr(S, Name)
+end
+
+const AbstractLabelledReactionNet = AbstractACSetType(TheoryLabelledReactionNet)
+const LabelledReactionNetUntyped = ACSetType(TheoryLabelledReactionNet, index=[:it,:is,:ot,:os])
+const LabelledReactionNet{R,C} = LabelledReactionNetUntyped{R,C,Symbol}
+
+LabelledReactionNet{R,C}(n, ts::Vararg{Union{Pair,Tuple}}) where {R,C} = begin
```

Comment:
I assume this function and everything preceding it is a copy-paste from AlgebraicPetri? I don't like that we have to copy-paste ~100 LoC just to test DPO for Petri nets.

We already have a solid set of tests in this module (good job!), so perhaps we should move these tests to AlgebraicPetri once the functionality is merged into Catlab. Thoughts @mehalter?

Reply:
I agree, the rewriting on Petri will get tested in the form of the notebooks @bosonbaas is building for the model space exploration. I think we can trust that the implementation is correct based on other schemas and then put these tests in AlgebraocPetri.